solr/core/src/java/org/apache/solr/response/transform/DeeplyNestedChildDocTransformer.java

```diff
@@ -0,0 +1,214 @@
+      final LeafReaderContext leafReaderContext = searcher.getIndexReader().leaves().get(seg);
+      final SortedDocValues segPathDocValues = DocValues.getSorted(leafReaderContext.reader(), NEST_PATH_FIELD_NAME);
+
+      Multimap<String,SolrDocument> pendingParentPathsToChildren = ArrayListMultimap.create();
+
+      if(children.matches() > 0) {
+        SolrDocumentFetcher docFetcher = searcher.getDocFetcher();
+        Set<String> dvFieldsToReturn = docFetcher.getNonStoredDVs(true).stream()
+            .filter(name -> !NEST_PATH_FIELD_NAME.equals(name)).collect(Collectors.toSet());
+        boolean shouldDecorateWithDVs = dvFieldsToReturn.size() > 0;
+        DocIterator i = children.iterator();
+        final int segBaseId = leafReaderContext.docBase;
+        final int firstChildDocId = i.nextDoc();
+        assert firstChildDocId < rootDocId;
+
+        for (int docId = firstChildDocId; docId < rootDocId; ++docId) {
+          // get the path
+          String fullDocPath = getPathByDocId(docId - segBaseId, segPathDocValues);
+
+          // Is this doc a direct ancestor of another doc we've seen?
+          boolean isAncestor = pendingParentPathsToChildren.containsKey(fullDocPath);
+
+          // Do we need to do anything with this doc (either ancestor or a matched the child query)
+          if (isAncestor || children.exists(docId)) {
+            // load the doc
+            SolrDocument doc = DocsStreamer.convertLuceneDocToSolrDoc(docFetcher.doc(docId),
+                schema, new SolrReturnFields());
+            doc.setField(NEST_PATH_FIELD_NAME, fullDocPath);
+            if (shouldDecorateWithDVs) {
+              docFetcher.decorateDocValueFields(doc, docId, dvFieldsToReturn);
+            }
+            // get parent path
+            // put into pending
+            String parentDocPath = lookupParentPath(fullDocPath);
+            pendingParentPathsToChildren.put(parentDocPath, doc); // multimap add (won't replace)
+
+            // if this path has pending child docs, add them.
+            if (isAncestor) {
+              addChildrenToParent(doc, pendingParentPathsToChildren.get(fullDocPath));
+              pendingParentPathsToChildren.removeAll(fullDocPath); // no longer pending
+            }
+          }
+        }
+
+        // only children of parent remain
+        assert pendingParentPathsToChildren.keySet().size() == 1;
+
+        addChildrenToParent(rootDoc, pendingParentPathsToChildren.get(null));
+      }
+    } catch (IOException e) {
+      rootDoc.put(getName(), "Could not fetch child Documents");
+    }
+  }
+
+  void addChildToParent(SolrDocument parent, SolrDocument child, String label) {
+    // lookup leaf key for these children using path
+    // depending on the label, add to the parent at the right key/label
+    // TODO: unfortunately this is the 2nd time we grab the paths for these docs. resolve how?
+    String trimmedPath = trimSuffixFromPaths(getLastPath(label));
```

Comment:
I'm a bit confused here.  Firstly, it's curious to see getLastPath(label) -- since label should not have a path if it is a label.  A label is a plain string key/label joining a parent to child like "comment".  Secondly, why the trimSuffixFromPaths... I don't get the point.  As I've said, comments of input/output sample values can help a ton in knowing what's going on.

Reply:
We don't want the # and trailing number to be added to the fieldName represented in the document hierarchy, so trimSuffixFromPath is used.

Reply:
Ok.  So I propose renaming the "label" variable to be "nestPath" since that's what it seems to be.  getLastPath is fine.  trimSuffixFromPaths should probable be renamed to trimLastChildIndex.  Then, "trimmedPath" variable would in fact be the "label" so should be named as such.